### PR TITLE
feat: create a training workflow that makes running experiments simple

### PR DESCRIPTION
**Pull request by atomwalk12**
The contribution is to devise a trainer to train, validate and easily keep track of future experiments

### CODE REVIEW

File: linalgzero/experiments/linalgzero_debug.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Training
+batch_size: 256
+train_iterations: 1000
+print_iterations: 100
+wandb_iterations: 500
+val_iterations: 500
```
[the remark this answers is not in the document]
Comment (atomwalk12) [reply linalgzero/experiments/linalgzero_debug.yml:6]: This is fine for now

File: tests/test_config.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def test_default_config():
+    """Test that the default config is valid."""
+    config_path = get_default_config_path()
+    parser = HfArgumentParser(ZeroConfig)
+
+    # This follows the "test the behaviour, not the implementation" principle.
+    # It will fail if the config has missing entries or the data class is defined incorrectly.
+    args: ZeroConfig = parser.parse_yaml_file(config_path)
```
[the remark this answers is not in the document]
Comment (atomwalk12) [reply tests/test_config.py:13]: Fixed

File: linalgzero/metrics/accuracy.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        """Update accuracy with a new batch of predictions and targets.
+
+        Expects `model_output` to have a 'logits' key and `batch` to have a 'labels' key.
+        """
+        logits = model_output["logits"]
+        labels = batch["label"]
+
+        pred = torch.argmax(logits.detach(), dim=-1)
+        if pred.shape != labels.shape:
+            raise IncompatibleShapesError(expected_shape=labels.shape, actual_shape=pred.shape)
```
[the remark this answers is not in the document]
Comment (atomwalk12) [reply linalgzero/metrics/accuracy.py:29]: Was added

File: linalgzero/trainers/trainer.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        self.global_step = 0
+        self.best_score: float = -float("inf")
+
+        # Setup
+        self.device = self._setup_device()
+        self.session_manager = SessionManager(config)
+        self.wandb_logger = WandbLogger(
+            config=config,
+            project_name=config.wandb_project,
+            run_name=self.session_manager.session_path.name,
```
[the remark this answers is not in the document]
Comment (atomwalk12) [reply linalgzero/trainers/trainer.py:45]: Fixed

File: linalgzero/utils/helpers.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    # Basic configuration
+    log_path.parent.mkdir(parents=True, exist_ok=True)
+    logging.basicConfig(
+        level=logging.INFO,
+        format="%(asctime)s [%(levelname)s] %(message)s",
+        handlers=[
+            logging.FileHandler(log_path),
+            logging.StreamHandler(),
+        ],
+    )
```
[the remark this answers is not in the document]
Comment (atomwalk12) [reply linalgzero/utils/helpers.py:54]: It is better if it fails instead of silently ignoring errors
